File: web-front-reactive/src/main/java/template/modules/admin/presentations/users/UserHandler.java
package template.modules.admin.presentations.users;

import com.undecided.primitive.list.Lists2;
import com.undecided.primitive.longs.Longs;
import com.undecided.primitive.misc.NodeIdGenerator;
import com.undecided.primitive.misc.SnowflakeIdGenerator;
import com.undecided.primitive.string.Strings2;
import lombok.AllArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.stereotype.Component;
import org.springframework.util.MultiValueMap;
import org.springframework.web.reactive.function.server.ServerRequest;
import org.springframework.web.reactive.function.server.ServerResponse;
import reactor.core.publisher.Mono;
import template.modules.admin.command.users.AddUserCommand;
import template.modules.admin.command.users.DeleteUserCommand;
import template.modules.admin.command.users.UpdateUserCommand;
import template.modules.admin.domain.users.User;
import template.modules.admin.presentations.ErrorMessage;
import template.modules.admin.query.users.UserFetcher;

@Component
@AllArgsConstructor
@Slf4j
public class UserHandler {
    private final AddUserCommand addUserCommand;
    private final UpdateUserCommand updateUserCommand;
    private final DeleteUserCommand deleteUserCommand;
    private final UserFetcher userFetcher;

    public Mono<ServerResponse> findAll(ServerRequest request) {
        return ServerResponse
                .ok()
                .contentType(MediaType.APPLICATION_JSON)
                .body(userFetcher.findAll(), User.class);
    }

    public Mono<ServerResponse> search(ServerRequest request) {
        MultiValueMap<String, String> criteria = request.queryParams();
        if (criteria.isEmpty()) {
            return ServerResponse
                    .badRequest()
                    .contentType(MediaType.APPLICATION_JSON)
                    .body(new ErrorMessage("Search must have query params"), ErrorMessage.class);
        }
        if (criteria.containsKey("email")) {
            return ServerResponse
                    .badRequest()
                    .contentType(MediaType.APPLICATION_JSON)
                    .body(new ErrorMessage("Search must have query params"), ErrorMessage.class);
        }

        var criteriaValue = Lists2.getFirst(criteria.get("email"));
        if (Strings2.isBlank(criteriaValue)) {
            return ServerResponse
                    .badRequest()
                    .contentType(MediaType.APPLICATION_JSON)
                    .body(new ErrorMessage("Incorrect search criteria value"), ErrorMessage.class);
        } else {
            return ServerResponse
                    .ok()
                    .contentType(MediaType.APPLICATION_JSON)
                    .body(userFetcher.findByEmail(criteriaValue), User.class);
        }

    }

    public Mono<ServerResponse> findUser(ServerRequest request) {
        Long id = Longs.toLongOrNull(request.pathVariable("id"));
        if (id == null) {
            return ServerResponse
                    .badRequest()
                    .contentType(MediaType.APPLICATION_JSON)
                    .body(new ErrorMessage("`id` must be numeric"), ErrorMessage.class);
        } else {
            var user = userFetcher.findById(id);
            if (user == null) {
                return ServerResponse
                        .notFound()
                        .build();
            } else {
                return ServerResponse
                        .ok()
                        .contentType(MediaType.APPLICATION_JSON)
                        .body(user, User.class);
            }
        }
    }

    public Mono<ServerResponse> addUser(ServerRequest request) {

        Mono<UserDto> newUser;
        try {
            newUser = request.bodyToMono(UserDto.class);
        } catch (Exception e) {
            log.error("Decoding body error", e);
            return ServerResponse
                    .status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .contentType(MediaType.APPLICATION_JSON)
                    .body(e, ErrorMessage.class);
        }
        if (newUser == null) {
            return ServerResponse
                    .badRequest()
                    .body(new ErrorMessage("Invalid body"), ErrorMessage.class);
        } else {
            Long id = new SnowflakeIdGenerator(NodeIdGenerator.generateNodeId()).generateID();
            User newUser2 = newUser.block().convertToModel(id);
            Mono<User> user = addUserCommand.execute(newUser2);
            if (user.block() == null) {
                return ServerResponse
                        .status(HttpStatus.INTERNAL_SERVER_ERROR)
                        .contentType(MediaType.APPLICATION_JSON)
                        .body(new ErrorMessage("Internal error"), ErrorMessage.class);
            } else {
                return ServerResponse
                        .status(HttpStatus.CREATED)
                        .contentType(MediaType.APPLICATION_JSON)
                        .body(user, User.class);
            }
        }
    }

    public Mono<ServerResponse> updateUser(ServerRequest request) {
        Long id = Longs.toLongOrNull(
                request.pathVariable("id"));

        if (id == null) {
            return ServerResponse
                    .badRequest()
                    .contentType(MediaType.APPLICATION_JSON)
                    .body(new ErrorMessage("`id` must be numeric"), ErrorMessage.class);
        } else {
            Mono<UserDto> updateUser;
            try {
                updateUser = request.bodyToMono(UserDto.class);
            } catch (Exception e) {
                log.error("Decoding body error", e);
                updateUser = null;
            }
            if (updateUser == null) {
                return ServerResponse
                        .badRequest()
                        .contentType(MediaType.APPLICATION_JSON)
                        .body(new ErrorMessage("Invalid body"), ErrorMessage.class);
            } else {

                Mono<User> user = updateUserCommand.execute(updateUser.block().convertToModel(id));
                if (user == null) {
                    return ServerResponse
                            .status(HttpStatus.NOT_FOUND)
                            .contentType(MediaType.APPLICATION_JSON)
                            .body(new ErrorMessage("Resource $id not found"), ErrorMessage.class);
                } else {
                    return ServerResponse
                            .ok()
                            .body(user, User.class);
                }
            }
        }
    }

    public Mono<ServerResponse> deleteUser(ServerRequest request) {
        var id = request.pathVariable("id");
        if (id == null) {
            return ServerResponse
                    .badRequest()
                    .body(new ErrorMessage("`id` must be numeric"), ErrorMessage.class);
        } else {
            if (deleteUserCommand.execute(Long.valueOf(id))) {
                return ServerResponse.noContent().build();
            } else {
                return ServerResponse
                        .notFound()
                        .build();
//                        .build(new ErrorMessage("Resource $id not found"),ErrorMessage.class);
            }
        }

    }
}

File: web-front-reactive2/src/main/kotlin/template/modules/admin/presentations/user/UserHandler.kt
package template.modules.admin.presentations.user

import kotlinx.coroutines.reactive.awaitFirstOrNull
import org.slf4j.LoggerFactory
import org.springframework.http.HttpStatus
import org.springframework.stereotype.Component
import org.springframework.web.reactive.function.server.*
import template.modules.admin.command.users.AddUserCommand
import template.modules.admin.command.users.DeleteUserCommand
import template.modules.admin.command.users.UpdateUserCommand
import template.modules.admin.query.users.UserFetcher

@Component
class UserHandler(
    private var addUserCommand: AddUserCommand,
    private var updateUserCommand: UpdateUserCommand,
    private var deleteUserCommand: DeleteUserCommand,
    private var userFetcher: UserFetcher
) {
    private val logger = LoggerFactory.getLogger(UserHandler::class.java)

    suspend fun findAll(request: ServerRequest): ServerResponse {
        val users = userFetcher.findAll()
        return ServerResponse.ok().json().bodyAndAwait(users)
    }

    suspend fun search(request: ServerRequest): ServerResponse {
        val criterias = request.queryParams()
        return when {
            criterias.isEmpty() -> ServerResponse.badRequest().json()
                .bodyValueAndAwait(ErrorMessage("Search must have query params"))

            criterias.contains("email") -> {
                val criteriaValue = criterias.getFirst("email")
                if (criteriaValue.isNullOrBlank()) {
                    ServerResponse.badRequest().json()
                        .bodyValueAndAwait(ErrorMessage("Incorrect search criteria value"))
                } else {
                    ServerResponse.ok().json().bodyAndAwait(userFetcher.findByEmail(criteriaValue))
                }
            }

            else -> ServerResponse.badRequest().json().bodyValueAndAwait(ErrorMessage("Incorrect search criteria"))
        }
    }

    suspend fun findUser(request: ServerRequest): ServerResponse {
        val id = request.pathVariable("id").toLongOrNull()
        return if (id == null) {
            ServerResponse.badRequest().json().bodyValueAndAwait(ErrorMessage("`id` must be numeric"))
        } else {
            val user = userFetcher.findById(id)
            if (user == null) ServerResponse.notFound().buildAndAwait()
            else ServerResponse.ok().json().bodyValueAndAwait(user)
        }
    }

    suspend fun addUser(request: ServerRequest): ServerResponse {
        val newUser = try {
            request.bodyToMono<UserDto>().awaitFirstOrNull()
        } catch (e: Exception) {
            logger.error("Decoding body error", e)
            null
        }
        return if (newUser == null) {
            ServerResponse.badRequest().json().bodyValueAndAwait(ErrorMessage("Invalid body"))
        } else {
            val user = addUserCommand.execute(newUser.toModel())
            if (user == null) ServerResponse.status(HttpStatus.INTERNAL_SERVER_ERROR).json()
                .bodyValueAndAwait(ErrorMessage("Internal error"))
            else ServerResponse.status(HttpStatus.CREATED).json().bodyValueAndAwait(user)
        }
    }

    suspend fun updateUser(request: ServerRequest): ServerResponse {
        val id = request.pathVariable("id").toLongOrNull()
        return if (id == null) {
            ServerResponse.badRequest().json().bodyValueAndAwait(ErrorMessage("`id` must be numeric"))
        } else {
            val updateUser = try {
                request.bodyToMono<UserDto>().awaitFirstOrNull()
            } catch (e: Exception) {
                logger.error("Decoding body error", e)
                null
            }
            if (updateUser == null) {
                ServerResponse.badRequest().json().bodyValueAndAwait(ErrorMessage("Invalid body"))
            } else {
                val user = updateUserCommand.execute(id, updateUser.toModel())
                if (user == null) ServerResponse.status(HttpStatus.NOT_FOUND).json()
                    .bodyValueAndAwait(ErrorMessage("Resource $id not found"))
                else ServerResponse.status(HttpStatus.OK).json().bodyValueAndAwait(user)
            }
        }
    }

    suspend fun deleteUser(request: ServerRequest): ServerResponse {
        val id = request.pathVariable("id").toLongOrNull()
        return if (id == null) {
            ServerResponse.badRequest().json().bodyValueAndAwait(ErrorMessage("`id` must be numeric"))
        } else {
            if (deleteUserCommand.execute(id)) ServerResponse.noContent().buildAndAwait()
            else ServerResponse.status(HttpStatus.NOT_FOUND).json()
                .bodyValueAndAwait(ErrorMessage("Resource $id not found"))
        }
    }
}